CommonMark/ExtensibleHtmlPrinter.cs

```diff
@@ -0,0 +1,742 @@
+            {
+                _htmlTextWriter.Write(buffer, lastPos - input.StartIndex, pos - lastPos);
+                lastPos = pos + 1;
+
+                switch (input.Source[pos])
+                {
+                    case '<':
+                        _htmlTextWriter.WriteConstant(EscapeHtmlLessThan);
+                        break;
+
+                    case '>':
+                        _htmlTextWriter.WriteConstant(EscapeHtmlGreaterThan);
+                        break;
+
+                    case '&':
+                        _htmlTextWriter.WriteConstant(EscapeHtmlAmpersand);
+                        break;
+
+                    case '"':
+                        _htmlTextWriter.WriteConstant(EscapeHtmlQuote);
+                        break;
+                }
+            }
+
+            _htmlTextWriter.Write(buffer, lastPos - input.StartIndex, input.Length - lastPos + input.StartIndex);
+        }
+
+        /// <summary>
+        /// Escapes special HTML characters.
+        /// </summary>
+        /// <remarks>Orig: escape_html(inp, preserve_entities)</remarks>
+        protected void EscapeHtml(StringContent inp)
+        {
+            int pos;
+            int lastPos;
+            char[] buffer = _htmlTextWriter.Buffer;
+
+            var parts = inp.RetrieveParts();
+            for (var i = parts.Offset; i < parts.Offset + parts.Count; i++)
+            {
+                var part = parts.Array[i];
+
+                if (buffer.Length < part.Length)
+                    buffer = _htmlTextWriter.Buffer = new char[part.Length];
+
+                part.Source.CopyTo(part.StartIndex, buffer, 0, part.Length);
+
+                lastPos = pos = part.StartIndex;
+                while ((pos = part.Source.IndexOfAny(EscapeHtmlCharacters, lastPos, part.Length - lastPos + part.StartIndex)) != -1)
+                {
+                    _htmlTextWriter.Write(buffer, lastPos - part.StartIndex, pos - lastPos);
+                    lastPos = pos + 1;
+
+                    switch (part.Source[pos])
+                    {
+                        case '<':
+                            _htmlTextWriter.WriteConstant(EscapeHtmlLessThan);
+                            break;
+
+                        case '>':
+                            _htmlTextWriter.WriteConstant(EscapeHtmlGreaterThan);
+                            break;
+
+                        case '&':
+                            _htmlTextWriter.WriteConstant(EscapeHtmlAmpersand);
+                            break;
+
+                        case '"':
+                            _htmlTextWriter.WriteConstant(EscapeHtmlQuote);
+                            break;
+                    }
+                }
+
+                _htmlTextWriter.Write(buffer, lastPos - part.StartIndex, part.Length - lastPos + part.StartIndex);
+            }
+        }
+
+        public void Print(TextWriter writer, Block block, CommonMarkSettings settings)
```

Comment:
`settings` should be saved in a protected read-only property so that overrides can read settings if there are any they need.